crates/page_table_entry/src/arch/loongarch64.rs

```diff
@@ -0,0 +1,161 @@
+//! loongarch64 page table entries.
+
+use crate::{GenericPTE, MappingFlags};
+use aarch64_cpu::registers::SPSR_EL1::M;
+use core::fmt;
+use memory_addr::PhysAddr;
+
+bitflags::bitflags! {
+    /// Page-table entry flags.
+    #[derive(Debug)]
+    pub struct PTEFlags: usize {
+        /// Whether the PTE is valid.
+        const V = 1 << 0;
+        /// Indicates the virtual page has been written since the last time the
+        /// D bit was cleared.
+        const D = 1 << 1;
+        /// Privilege Level with 2 bits.
+        const PLVL = 1 << 2;
+        const PLVH = 1 << 3;
+        /// Memory Access Type controls the type of access, such as whether it
+        /// can be cached by Cache, etc.
+        const MATL = 1 << 4;
+        const MATH = 1 << 5;
+        /// Designates a global mapping OR Whether the page is huge page.
+        const GH = 1 << 6;
+        /// Whether the physical page is exist.
+        const P = 1 << 7;
+        /// Whether the page is writable.
+        const W = 1 << 8;
+        /// Designates a global mapping when using huge page.
+        const G = 1 << 12;
+        /// Whether the page is not readable.
+        const NR = 1 << 61;
+        /// Whether the page is not executable.
+        const NX = 1 << 62;
+        /// Whether the privilege Level is restricted. When RPLV is 0, the PTE
+        /// can be accessed by any program with privilege Level higher than PLV.
+        const RPLV = 1 << 63;
+    }
+}
+
+impl From<PTEFlags> for MappingFlags {
+    fn from(f: PTEFlags) -> Self {
+        let mut ret = Self::empty();
+        if !f.contains(PTEFlags::NR) {
+            ret |= Self::READ;
+        }
+        if f.contains(PTEFlags::W) {
+            ret |= Self::WRITE;
+        }
+        if !f.contains(PTEFlags::NX) {
+            ret |= Self::EXECUTE;
+        }
+        if f.contains(PTEFlags::PLVL | PTEFlags::PLVH) {
+            ret |= Self::USER;
+        }
+        if f.contains(PTEFlags::MATH) {
+            // MAT = 2 (Weakly-ordered UnCached) -> UNCACHED
+            ret |= Self::UNCACHED;
+        } else if !f.contains(PTEFlags::MATL) {
+            // MAT = 0 (Strongly-ordered UnCached) -> DEVICE
+            ret |= Self::DEVICE;
+        }
+        ret
+    }
+}
+
+impl From<MappingFlags> for PTEFlags {
+    fn from(f: MappingFlags) -> Self {
+        if f.is_empty() {
+            return Self::empty();
+        }
+        let mut ret = Self::V;
+        if !f.contains(MappingFlags::READ) {
+            ret |= Self::NR;
+        }
+        if f.contains(MappingFlags::WRITE) {
+            ret |= Self::W;
+        }
+        if !f.contains(MappingFlags::EXECUTE) {
+            ret |= Self::NX;
+        }
+        if f.contains(MappingFlags::USER) {
+            ret |= Self::PLVH | Self::PLVL;
+        }
+        if !f.contains(MappingFlags::DEVICE | MappingFlags::UNCACHED) {

```

Comment:
`!A.contains(B | C)` != `!A.contains(B) && !A.contains(C)`, so this expression is wrong.

You can change it to the following with better readability:

```rust
if !f.contains(MappingFlags::DEVICE) {
  if f.contains(MappingFlags::UNCACHED) {
    // MAT = 2 (WUC)
    ret |= Self::MATH
  } else {
    // MAT = 1 (Coherent Cached)
    ret |= Self::MATL
  }
} // else, MAT = 0 (SUC)
```

Reply:
Yes, this is a stupid mistake